src/controllers/trustpub/tokens/exchange/mod.rs

```diff
@@ -0,0 +1,146 @@
+use super::json;
+use crate::app::AppState;
+use crate::util::errors::{AppResult, bad_request, server_error};
+use axum::Json;
+use crates_io_database::models::trustpub::{NewToken, NewUsedJti};
+use crates_io_database::schema::trustpub_configs_github;
+use crates_io_diesel_helpers::lower;
+use crates_io_trustpub::access_token::AccessToken;
+use crates_io_trustpub::github::{GITHUB_ISSUER_URL, GitHubClaims};
+use crates_io_trustpub::unverified::UnverifiedClaims;
+use diesel::prelude::*;
+use diesel::result::DatabaseErrorKind::UniqueViolation;
+use diesel::result::Error::DatabaseError;
+use diesel_async::scoped_futures::ScopedFutureExt;
+use diesel_async::{AsyncConnection, RunQueryDsl};
+use secrecy::ExposeSecret;
+
+#[cfg(test)]
+mod tests;
+
+/// Exchange an OIDC token for a temporary access token.
+#[utoipa::path(
+    put,
+    path = "/api/v1/trusted_publishing/tokens",
+    request_body = inline(json::ExchangeRequest),
+    tag = "trusted_publishing",
+    responses((status = 200, description = "Successful Response", body = inline(json::ExchangeResponse))),
+)]
+pub async fn exchange_trustpub_token(
+    state: AppState,
+    json: json::ExchangeRequest,
+) -> AppResult<Json<json::ExchangeResponse>> {
```

Comment:
in case you didn't find it in the large diff: this is the main entrypoint of the new API endpoint 😉